From: Andi Kleen <ak@linux.intel.com>
Date: Fri, 25 Feb 2011 14:44:28 -0800
Subject: [PATCH] thp: fix interleaving for transparent hugepages
References: THP core (fate #311931)
Git-commit: 8eac563c1c3a2047083022357ae63722b19e4e08
Patch-mainline: v2.6.38-rc7

The THP code didn't pass the correct interleaving shift to the memory
policy code.  Fix this here by adjusting for the order.

Signed-off-by: Andi Kleen <ak@linux.intel.com>
Reviewed-by: Christoph Lameter <cl@linux.com>
Acked-by: Andrea Arcangeli <aarcange@redhat.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>
Signed-off-by: Mel Gorman <mgorman@suse.de>
---
 mm/mempolicy.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/mm/mempolicy.c b/mm/mempolicy.c
index c6018b0..1425a5d 100644
--- a/mm/mempolicy.c
+++ b/mm/mempolicy.c
@@ -1623,7 +1623,7 @@ alloc_pages_vma(gfp_t gfp, int order, struct vm_area_struct *vma,
 	if (unlikely(pol->mode == MPOL_INTERLEAVE)) {
 		unsigned nid;
 
-		nid = interleave_nid(pol, vma, addr, PAGE_SHIFT);
+		nid = interleave_nid(pol, vma, addr, PAGE_SHIFT + order);
 		mpol_cond_put(pol);
 		return alloc_page_interleave(gfp, order, nid);
 	}

